enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -361,70 +367,39 @@ func (np *nodePool) FindConnectedExecutorByID(executorID string) *executionNode
 	return nil
 }
 
-// unclaimedTasksList maintains a subset of unclaimed task IDs that can be given out to newly registered executors.
-// Only the most recent maxUnclaimedTasksTracked task IDs are kept.
-type unclaimedTasksList struct {
-	taskList *list.List
-	taskMap  map[string]*list.Element
-	mu       sync.Mutex
-}
-
-func newUnclaimedTasksList() *unclaimedTasksList {
-	l := &unclaimedTasksList{}
-	l.taskList = list.New()
-	l.taskMap = make(map[string]*list.Element)
-	return l
-}
-
-func (l *unclaimedTasksList) addTask(taskID string) {
-	l.mu.Lock()
-	defer l.mu.Unlock()
-	if _, exists := l.taskMap[taskID]; exists {
-		return
+func (np *nodePool) AddUnclaimedTask(ctx context.Context, taskID string) error {
+	key := np.key.redisUnclaimedTasksKey()
+	m := &redis.Z{
+		Member: taskID,
+		Score:  float64(time.Now().Unix()),
 	}
-	e := l.taskList.PushBack(taskID)
-	l.taskMap[taskID] = e
-	if l.taskList.Len() > maxUnclaimedTasksTracked {
-		v := l.taskList.Remove(l.taskList.Front())
-		s, ok := v.(string)
-		if !ok { // Should never happen.
-			log.Warningf("non-string value in list: %T", v)
-			return
-		}
-		delete(l.taskMap, s)
+	err := np.rdb.ZAdd(ctx, key, m).Err()
+	if err != nil {
+		return err
+	}
+	err = np.rdb.Expire(ctx, key, unclaimedTaskSetTTL).Err()
+	if err != nil {
+		return err
 	}
-}
 
-func (l *unclaimedTasksList) removeTask(taskID string) {
-	l.mu.Lock()
-	defer l.mu.Unlock()
-	e, ok := l.taskMap[taskID]
-	if !ok {
-		return
+	// Trim the set if necessary.
+	// The next 2 commands are not atomic but it's okay if the list length is not exactly what we want.
+	n, err := np.rdb.ZCard(ctx, key).Result()
+	if err != nil {
+		return err
 	}
-	delete(l.taskMap, taskID)
-	l.taskList.Remove(e)
+	if n > maxUnclaimedTasksTracked {
+		return np.rdb.ZRemRangeByRank(ctx, key, 0, n-maxUnclaimedTasksTracked-1).Err()
+	}
```

Comment:
lol not afraid to admit i had to think about this for way too long to figure it out. Could you add a comment that explains what ZREMRANGEBYRANK is doing and also explains the n-maxTasks-1 math?

Reply:
Done.